server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1098,37 +1109,59 @@ SinkSchemaAnnouncer getSinkSchemaAnnouncer()
   /**
    * Unannounces the given base segment and all its upgraded versions.
    */
-  private void unannounceAllVersionsOfSegment(DataSegment baseSegment) throws IOException
+  private void unannounceAllVersionsOfSegment(DataSegment baseSegment, Sink sink)
   {
-    segmentAnnouncer.unannounceSegment(baseSegment);
+    synchronized (sink) {
+      final SegmentIdWithShardSpec baseId = SegmentIdWithShardSpec.fromDataSegment(baseSegment);
+      if (!baseSegmentToUpgradedSegments.containsKey(baseId)) {
+        return;
+      }
 
-    final Set<SegmentIdWithShardSpec> upgradedVersionsOfSegment
-        = baseSegmentToUpgradedVersions.remove(baseSegment.getId());
-    if (upgradedVersionsOfSegment == null || upgradedVersionsOfSegment.isEmpty()) {
-      return;
+      final List<SegmentIdWithShardSpec> upgradedVersionsOfSegment
+          = ImmutableList.copyOf(baseSegmentToUpgradedSegments.get(baseId));
+
+      for (SegmentIdWithShardSpec newId : upgradedVersionsOfSegment) {
+        final DataSegment newSegment = new DataSegment(
+            newId.getDataSource(),
+            newId.getInterval(),
+            newId.getVersion(),
+            baseSegment.getLoadSpec(),
+            baseSegment.getDimensions(),
+            baseSegment.getMetrics(),
+            newId.getShardSpec(),
+            baseSegment.getBinaryVersion(),
+            baseSegment.getSize()
+        );
+        unannounceSegment(newSegment);
+      }
     }
+  }
 
-    for (SegmentIdWithShardSpec newId : upgradedVersionsOfSegment) {
-      final DataSegment newSegment = new DataSegment(
-          newId.getDataSource(),
-          newId.getInterval(),
-          newId.getVersion(),
-          baseSegment.getLoadSpec(),
-          baseSegment.getDimensions(),
-          baseSegment.getMetrics(),
-          newId.getShardSpec(),
-          baseSegment.getBinaryVersion(),
-          baseSegment.getSize()
-      );
-      segmentAnnouncer.unannounceSegment(newSegment);
+  private void unannounceSegment(DataSegment segment)
+  {
+    try {
+      segmentAnnouncer.unannounceSegment(segment);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Failed to unannounce segment[%s]", schema.getDataSource())
+         .addData("identifier", segment.getId().toString())
+         .emit();
+    }
+    SegmentIdWithShardSpec id = SegmentIdWithShardSpec.fromDataSegment(segment);
+    SegmentIdWithShardSpec baseId = upgradedSegmentToBaseSegment.remove(id);
+    if (baseId == null) {
+      return;
+    }
+    baseSegmentToUpgradedSegments.get(baseId).remove(id);
```

Comment:
Even though the set here cannot be null, it is best to handle the null case too.

Reply:
This no longer needs to be handled because of https://github.com/apache/druid/pull/16162#discussion_r1579326534